<launch>
  <arg name="INPUT_IMAGE" default="/camera_remote/rgb/image_raw" />
  <arg name="INPUT_CLOUD" default="/camera_remote/aligned_depth_to_color/points" />

  <arg name="BGR_MASK_IMAGE" default="/cutie_node/output/segmentation_bgr" />
  <arg name="MASK_IMAGE" default="/cutie_node/output/segmentation" />

  <arg name="standalone" default="true"/> <!-- debug nodes by setting then standalone -->
  <arg name="MANAGER" value="detic_detection_manager" unless="$(arg standalone)"/>
  <arg name="MANAGER" value="" if="$(arg standalone)"/>
  <arg name="LOAD_STATEMENT" value="load" unless="$(arg standalone)"/>
  <arg name="LOAD_STATEMENT" value="standalone" if="$(arg standalone)"/>
  <arg name="target_frame_id" default="camera_link" />

  <arg name="run_rviz" default="true"/>
  <arg name="run_collection" default="true"/>

  <arg name="object_frame_name" default="objectoutput00"/>

  <node name="$(arg MANAGER)" pkg="nodelet" type="nodelet" args="manager"/>

  <node name="image_format_converter" pkg="jsk_2023_09_cook_from_recipe" type="image_32sc1_to_bgr_converter.py" >
    <remap from="~input" to="$(arg MASK_IMAGE)" />
    <remap from="~output" to="$(arg BGR_MASK_IMAGE)" />
  </node>

  <node name="apply_mask_image" pkg="jsk_perception" type="apply_mask_image" clear_params="true">
    <remap from="~input" to="$(arg INPUT_IMAGE)" />
    <remap from="~input/mask" to="$(arg BGR_MASK_IMAGE)" />
    <!-- <rosparam> -->
    <!--   approximate_sync: true -->
    <!-- </rosparam> -->
  </node>

  <node name="detic_label_image_to_indices"
        pkg="nodelet" type="nodelet"
        args="$(arg LOAD_STATEMENT) jsk_pcl_utils/LabelToClusterPointIndices $(arg MANAGER)">
    <remap from="~input" to="$(arg MASK_IMAGE)"/>
    <remap from="~output" to="detic_segmentor/indices"/>
  </node>

  <!-- cluster_filter: 1 is desirable, but only stable with jsk_recognition/#2739 -->
  <node name="detic_euclidean_clustering"
        pkg="nodelet" type="nodelet"
        args="$(arg LOAD_STATEMENT) jsk_pcl/EuclideanClustering $(arg MANAGER)"
        clear_params="true">
    <remap from="~input" to="$(arg INPUT_CLOUD)"/>
    <remap from="~input/cluster_indices" to="detic_segmentor/indices"/>
    <!-- <rosparam> -->
    <!--   multi: true -->
    <!--   tolerance: 0.03 -->
    <!--   min_size: 10 -->
    <!--   downsample_enable: true -->
    <!--   approximate_sync: true -->
    <!--   queue_size: 100 -->
    <!-- </rosparam> -->
    <rosparam>
      multi: true
      tolerance: 0.03
      min_size: 10
      downsample_enable: true
      approximate_sync: true
      queue_size: 100
      cluster_filter: 1
    </rosparam>
  </node>

  <node name="detic_cluster_point_indices_decomposer"
        pkg="nodelet" type="nodelet"
        args="$(arg LOAD_STATEMENT) jsk_pcl/ClusterPointIndicesDecomposer $(arg MANAGER)"
        clear_params="true">
    <remap from="~input" to="$(arg INPUT_CLOUD)"/>
    <remap from="~target" to="detic_euclidean_clustering/output"/>
    <remap from="~boxes" to="detic_segmentor/output/boxes"/>
    <remap from="~centroid_pose_array" to="detic_segmentor/output/centroid"/>
    <!-- <rosparam subst_value="true"> -->
    <!--   align_boxes: true -->
    <!--   align_boxes_with_plane: false -->
    <!--   force_to_flip_z_axis: false -->
    <!--   use_pca: false -->
    <!--   target_frame_id: $(arg target_frame_id) -->
    <!--   approximate_sync: true -->
    <!--   queue_size: 100 -->
    <!-- </rosparam> -->
    <rosparam subst_value="true">
      align_boxes: true
      align_boxes_with_plane: false
      force_to_flip_z_axis: false
      use_pca: true
      target_frame_id: $(arg target_frame_id)
      approximate_sync: true
      queue_size: 100
      publish_tf: true
      tf_prefix: object
    </rosparam>
  </node>

  <node name="cluster_point_indices_to_point_indices"
        pkg="nodelet" type="nodelet"
        args="$(arg LOAD_STATEMENT) jsk_pcl_utils/ClusterPointIndicesToPointIndices $(arg MANAGER)"
        clear_params="true">
    <remap from="~input" to="detic_segmentor/indices"/>
    <rosparam>
      index: 0
    </rosparam>
  </node>

  <node name="extract_indices"
        pkg="nodelet" type="nodelet"
        args="$(arg LOAD_STATEMENT) jsk_pcl/ExtractIndices $(arg MANAGER)">
    <remap from="~input" to="$(arg INPUT_CLOUD)" />
    <remap from="~indices" to="cluster_point_indices_to_point_indices/output" />
    <rosparam>
      keep_organized: true
      approximate_sync: true
    </rosparam>
  </node>

  <node name="point_cloud_concat" pkg="nodelet" type="nodelet"
        args="$(arg LOAD_STATEMENT) pcl/PointCloudConcatenateDataSynchronizer $(arg MANAGER)" output="screen">
    <rosparam subst_value="true">
      approximate_sync: true
      queue_size: 10
      output_frame: $(arg object_frame_name)
      input_topics:
      - /extract_indices/output
      - /extract_indices/output
    </rosparam>
  </node>

  <include file="$(find jsk_2023_09_cook_from_recipe)/launch/tracking_data_collection.launch" if="$(arg run_collection)" />

  <!-- rviz -->
  <group if="$(arg run_rviz)">
    <node name="rviz"
          pkg="rviz" type="rviz"
          args="-d $(find jsk_2023_09_cook_from_recipe)/rviz/rs_apply_track_object_mask.rviz">
    </node>
  </group>

</launch>
